arc: add XY mem support

SiyuanCheng-CN opened the pull request:
## Introduction
The XY Memory is a feature commonly found in DSP processors to increase the DSP performance. The XY component allows a ARC processor to implicitly load source operands and store results into a closely coupled memory using a single instruction. This results in dense code and high performance. 


## Proposed change
Add XY memory for ARC EM9D/EM11D processors including em_starterkit, em_starterkit_em11d. emsdp_em9d, nsim_em, iotdk.
Add XY memory into arc mpu region.

galak commented on dts/bindings/arc/arc,xccm.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2017, synopsy
```
copyright doesn't look correct, year, and name.

SiyuanCheng-CN replied:
Done, thanks for your check.

galak commented on dts/bindings/arc/arc,yccm.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2017, synopsy
```
copyright doesn't look correct, year, and name.

galak commented on dts/bindings/arc/arc,yccm.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2017, synopsy
+# SPDX-License-Identifier: Apache-2.0
+
+description: ARC YCCM
```
Can you add a little more detail to description.

galak commented on dts/bindings/arc/arc,xccm.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2017, synopsy
+# SPDX-License-Identifier: Apache-2.0
+
+description: ARC XCCM
```
Can you add a little more detail to description.

galak commented on boards/arc/nsim/nsim.dtsi:
```diff
@@ -9,6 +9,22 @@
 #include "skeleton.dtsi"
 #include <mem.h>
 
+#ifndef ICCM_ADDR
```
Where are these ifdef defines getting defined?

galak commented on boards/arc/nsim/nsim_em.dts:
```diff
@@ -8,9 +8,35 @@
 
 #include "nsim_em.dtsi"
 
+#ifndef XCCM_ADDR
```
Where are these ifdef defines getting defined?

SiyuanCheng-CN replied:
I followed the style of other memory, such as [ICCM and DCCM](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arc/nsim/nsim-ccm-mem.dtsi#:~:text=%23ifndef%20ICCM_ADDR,%23endif). Use #ifdef is to have different definition for special application, just like [arc_sec_application](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/arc_secure_services/nsim_sem_normal.dts#:~:text=%23define%20ICCM_ADDR%2040000,DCCM_SIZE%20DT_SIZE_K(256)).

fabiobaltieri replied:
That does not feel like a correct use of dts preprocessor macro, why not delete the existing node and redefine it how it's supposed to be instead? That way you wyould not have to do quirky things like `/* must be before include of nsim_em.dtsi */`. Not saying it should be fixed in this PR but maybe this should avoid copying that pattern (and then that can be fixed separately).

SiyuanCheng-CN replied:
removed `#ifndef` to avoid mislead

galak reviewed (changes requested):
dts bindings need fixing and the #defines in .dtsi files don't seem correct in that where are those defines getting set?

evgeniy-paltsev commented:
Hi @SiyuanCheng-CN ,

as I can see in one commit you are removing ICCM_ADDR, ICCM_SIZE, DCCM_ADDR, DCCM_SIZE definitions and in second commit you are bringing them back. I guess you need to remove that part from both commits.


PS: I'm talking about removing and adding that part:
```
#ifndef ICCM_ADDR
#define ICCM_ADDR 0
#endif

#ifndef ICCM_SIZE
#define ICCM_SIZE DT_SIZE_K(1024)
#endif

#ifndef DCCM_ADDR
#define DCCM_ADDR 80000000
#endif

#ifndef DCCM_SIZE
#define DCCM_SIZE DT_SIZE_K(1024)
#endif
```

SiyuanCheng-CN commented:
> Hi @SiyuanCheng-CN ,
> 
> as I can see in one commit you are removing ICCM_ADDR, ICCM_SIZE, DCCM_ADDR, DCCM_SIZE definitions and in second commit you are bringing them back. I guess you need to remove that part from both commits.
> 
> PS: I'm talking about removing and adding that part in `boards/arc/nsim/nsim.dtsi`:
> 
> ```
> #ifndef ICCM_ADDR
> #define ICCM_ADDR 0
> #endif
> 
> #ifndef ICCM_SIZE
> #define ICCM_SIZE DT_SIZE_K(1024)
> #endif
> 
> #ifndef DCCM_ADDR
> #define DCCM_ADDR 80000000
> #endif
> 
> #ifndef DCCM_SIZE
> #define DCCM_SIZE DT_SIZE_K(1024)
> #endif
> ```

mbolivar-nordic commented:
We can't merge this with the merge commit; please do a clean rebase

evgeniy-paltsev reviewed (changes requested):
Hi @SiyuanCheng-CN,

Le's squash `arc: optimize XY mem dts definition` commit into previous one as it's only fixup - so we don't want to have it as a separate commit.

And let's add a bit more info into the original `arc: add XY mem support` commit about what we do and why (at least please copy introduction section from this PR to commit message)

Thanks!

evgeniy-paltsev commented on boards/arc/nsim/nsim_em.dts:
```diff
@@ -8,9 +8,24 @@
 / {
 	model = "snps,nsim_em";
 	compatible = "snps,nsim_em";
+
+	xccm0: xccm@XCCM_ADDR {
+		compatible = "arc,xccm";
+		reg = <DT_ADDR(XCCM_ADDR) XCCM_SIZE>;
+	};
+
+	yccm0: yccm@YCCM_ADDR {
+		compatible = "arc,yccm";
+		reg = <DT_ADDR(YCCM_ADDR) YCCM_SIZE>;
+	};
```
Given that we don't actually use XY on `nsim_em` platform and we add `nsim_em11d` in DSP PR for DSP and XY testing - I'd propose to avoid unnecessary `nsim_em` platform changes.

Same for boards/arc/nsim/support/nsim_em.props & boards/arc/nsim/support/mdb_em.args files.

evgeniy-paltsev commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -62,6 +62,12 @@ MEMORY {
 	DCCM  (rw)  : ORIGIN = DCCM_START,  LENGTH = DCCM_SIZE
+#endif
+#ifdef XCCM_START
+	XCCM  (rw)  : ORIGIN = XCCM_START, 	LENGTH = XCCM_SIZE
```
```suggestion
	XCCM  (rw)  : ORIGIN = XCCM_START,  LENGTH = XCCM_SIZE
```

ruuddw commented on boards/arc/iotdk/arc_mpu_regions.c:
```diff
@@ -19,6 +19,20 @@ static struct arc_mpu_region mpu_regions[] = {
+	MPU_REGION_ENTRY("XCCM",
+			 DT_REG_ADDR(DT_INST(0, arc_xccm)),
+			 DT_REG_SIZE(DT_INST(0, arc_xccm)),
+			 REGION_KERNEL_RAM_ATTR | REGION_DYNAMIC),
```
iotdk has an MPU that supports overlapping regions, I don't think REGION_DYNAMIC is required (DCCM doesn't have the attribute either).
For the other boards, I can see that access may be restricted to certain tasks only, so dynamic allocation and switching access rights in context switches would be required.

Commit:
arc: add XY mem support

The XY Memory is a feature commonly found in DSP processors to increase
the DSP performance. The XY component allows a ARC processor to
implicitly load source operands and store results into a closely coupled
memory using a single instruction.

Add XY memory for ARC EM9D/EM11D processors including em_starterkit,
em_starterkit_em11d. emsdp_em9d, nsim_em, iotdk.

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

evgeniy-paltsev commented:
This PR has only arc-related changes and it has all approvals from ARC developers. As we don't have any additional comments we can finally merge this PR.

evgeniy-paltsev commented:
@fabiobaltieri I guess this can be merged now.

evgeniy-paltsev commented:
@galak may I ask you re-review this? This PR is waiting only your review now.

Thanks!
